File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/employee.dart
import 'package:equatable/equatable.dart';
import 'outlet_access.dart';

class Employee extends Equatable {
  final int id;
  final String name;
  final String username;
  final String? email;
  final String? phone;
  final String? gender;
  final String? formattedGender;
  final String? address;
  final int? age;
  final String? startDate;
  final String? dateOfBirth;
  final bool isActive;
  final int? outletId;
  final int? cutoffDays;
  final String? lastLoginAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final DateTime? deletedAt;
  final List<OutletAccess>? accessibleOutlets;

  const Employee({
    required this.id,
    required this.name,
    required this.username,
    this.email,
    this.phone,
    this.gender,
    this.formattedGender,
    this.address,
    this.age,
    this.startDate,
    this.dateOfBirth,
    required this.isActive,
    this.outletId,
    this.cutoffDays,
    this.lastLoginAt,
    this.createdAt,
    this.updatedAt,
    this.deletedAt,
    this.accessibleOutlets,
  });

  factory Employee.fromModel(dynamic model) {
    return Employee(
      id: model.id,
      name: model.name,
      username: model.username,
      email: model.email,
      phone: model.phone,
      gender: model.gender,
      formattedGender: model.formattedGender,
      address: model.address,
      age: model.age,
      startDate: model.startDate,
      dateOfBirth: model.dateOfBirth,
      isActive: model.isActive,
      outletId: model.outletId,
      cutoffDays: model.cutoffDays,
      lastLoginAt: model.lastLoginAt,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      deletedAt: model.deletedAt,
      accessibleOutlets: model.accessibleOutlets?.map<OutletAccess>((e) => e.toEntity()).toList(),
    );
  }

  @override
  List<Object?> get props => [
    id,
    name,
    username,
    email,
    phone,
    gender,
    formattedGender,
    address,
    age,
    startDate,
    dateOfBirth,
    isActive,
    outletId,
    cutoffDays,
    lastLoginAt,
    createdAt,
    updatedAt,
    deletedAt,
    accessibleOutlets,
  ];
}

File: laundry/wash_wallet_be/resources/js/types/employee.ts
import {
    Attendance,
    BaseFilters,
    EmployeePosition,
    EmployeeProcess,
    EmployeeSalary,
    Expense,
    FineLog,
    Loan,
    Order,
    Outlet,
    Payroll,
} from ".";

export interface Employee {
    id: number;
    outletId: number;
    name: string;
    username: string;
    email?: string | null;
    phone?: string;
    gender: "male" | "female";
    dateOfBirth?: string | null;
    avatar?: string | null;
    address?: string | null;
    startDate: string;
    isActive: boolean;
    cutoffDays: number;
    age?: number | null;
    lastLoginAt?: string | null;
    attendances: Attendance[];
    employeePositions?: EmployeePosition[] | null;
    employeeProcesses?: EmployeeProcess[] | null;
    employeeSalaries: EmployeeSalary[];
    expenses?: Expense[] | null;
    fineLogs: FineLog[];
    loans?: Loan[] | null;
    outlet: Outlet;
    orders?: Order[] | null;
    payrolls: Payroll[];
    employeeProcessesCount?: number;
    employeePositionsCount?: number;
    employeeSalariesCount?: number;
    isEligibleForProduction?: boolean;
    fineLogsCount?: number;
    ordersCount?: number;
    loansCount?: number;
    payrollsCount?: number;
    finesCount?: number;
    expensesCount?: number;
    attendancesCount?: number;
    createdAt: string;
    updatedAt: string;
}

export interface EmployeeCreateFormData {
    outletId: number | undefined;
    name: string;
    username: string;
    password: string;
    passwordConfirmation: string;
    avatar?: File | null;
    phone: string;
    address: string;
    dateOfBirth?: string;
    gender: "male" | "female" | "";
    startDate: string;
    isActive?: boolean;
    cutoffDays: number;
    positionIds?: number[];
    employeeSalaries?: EmployeeSalaryFormItem[];
    employeeProcesses?: EmployeeProcessFormItem[];
    employeeProcessCommissions?: EmployeeProcessCommissionFormItem[];
    [key: string]: any;
}

export interface EmployeeEditFormData {
    name: string;
    username: string;
    avatar?: File | null;
    phone: string;
    address: string;
    gender: "male" | "female" | "";
    startDate: string;
    isActive: boolean;
    cutoffDays: number;
    positionIds?: number[];
    employeeSalaries?: EmployeeSalaryFormItem[];
    employeeProcessCommissions?: EmployeeProcessCommissionFormItem[];
    [key: string]: any;
}

export interface EmployeeSalaryFormItem {
    salaryId: number | null;
    amount: number;
    [key: string]: any;
}

export interface EmployeeProcessFormItem {
    processId: number;
    isActive: boolean;
    [key: string]: any;
}

export interface EmployeeProcessCommissionFormItem {
    processId: number | null;
    commissionType: "per_item" | "per_kg" | "percentage" | "flat";
    commissionValue: number;
    hasTarget: boolean;
    targetThreshold: number;
    bonusAmount: number;
    effectiveDate?: string | null;
    [key: string]: any;
}

export interface EmployeeFilters extends BaseFilters {
    outletId?: number;
    positionId?: number;
    isActive?: "true" | "false";
    gender?: "male" | "female";
}

export interface EmployeeSalaryFormData {
    amount: number;
    salaryId: number;
    [key: string]: any;
}

export interface EmployeePositionFormData {
    positionId: number;
    isActive: boolean;
}

export interface EmployeeProcessCommissionFormData {
    processId: number;
    commissionType: "per_item" | "per_kg" | "percentage" | "flat";
    commissionValue: number;
    hasTarget: boolean;
    targetThreshold?: number;
    bonusAmount?: number;
    effectiveDate?: string | null;
    isActive: boolean;
    [key: string]: any;
}

export interface EmployeeProcessCreateFormData {
    processId: number;
    hasCommission: boolean;
    commissionType: "per_item" | "per_kg" | "percentage" | "flat";
    commissionValue: number;
    hasTarget: boolean;
    targetThreshold: number;
    bonusAmount: number;
}

export interface EmployeeLoanFormData {
    sourceAccountId: number | "";
    amount: number | "";
    installmentAmount: number | "";
    loanDate: string;
    dueDate?: string;
    status: "ongoing" | "paid" | "bad_debt";
    note?: string;
    [key: string]: any;
}
